oximeter/db/src/client.rs

```diff
@@ -1443,6 +1461,102 @@ mod tests {
+        let mut db = ClickHouseInstance::new(0)
+            .await
+            .expect("Failed to start ClickHouse");
+        let address = SocketAddr::new("::1".parse().unwrap(), db.port());
+        let log = Logger::root(slog::Discard, o!());
+        let client = Client::new(address, &log);
+        client
+            .init_db()
+            .await
+            .expect("Failed to initialize timeseries database");
+        client
+            .insert_samples(&samples)
+            .await
+            .expect("Failed to insert samples");
+        let timeseries_name = "service:request_latency";
+
+        // First, query without an order. We should see all the results in ascending order.
+        let all_measurements = &client
+            .select_timeseries_with(
+                timeseries_name,
+                &[],
+                None,
+                None,
+                None,
+                None,
+            )
+            .await
+            .expect("Failed to select timeseries")[0]
+            .measurements;
+
+        assert!(
+            all_measurements.len() > 1,
+            "need more than one measurement to test ordering"
+        );
+
+        // Explicitly specifying asc should give the same results
+        let timeseries_asc = &client
+            .select_timeseries_with(
+                timeseries_name,
+                &[],
+                None,
+                None,
+                None,
+                Some(PaginationOrder::Ascending),
+            )
+            .await
+            .expect("Failed to select timeseries")[0]
+            .measurements;
+        assert_eq!(all_measurements, timeseries_asc);
+
+        // Now get the results in reverse order
+        let timeseries_desc = &client
+            .select_timeseries_with(
+                timeseries_name,
+                &[],
+                None,
+                None,
+                None,
+                Some(PaginationOrder::Descending),
+            )
+            .await
+            .expect("Failed to select timeseries")[0]
+            .measurements;
+
+        let mut timeseries_asc_rev = timeseries_asc.clone();
+        timeseries_asc_rev.reverse();
+
+        assert_ne!(timeseries_desc, timeseries_asc);
+        assert_eq!(timeseries_desc, &timeseries_asc_rev);
+
+        // can use limit 1 to get single most recent measurement
+        let desc_limit_1 = &client
+            .select_timeseries_with(
+                timeseries_name,
+                &[],
+                None,
+                None,
+                Some(NonZeroU32::new(1).unwrap()),
+                Some(PaginationOrder::Descending),
+            )
+            .await
+            .expect("Failed to select timeseries")[0]
+            .measurements;
+
+        assert_eq!(desc_limit_1.len(), 1);
+        assert_eq!(
+            desc_limit_1.first().unwrap(),
+            timeseries_asc.last().unwrap()
+        );
```

Comment:
Is it just returning the last moment in time even if that moment doesn't have a value associated with it?

Reply:
`setup_select_test` produces more than one timeseries. Your call to `select_timeseries_with` does not provide any filtering criteria for the timeseries, so they're all returned. I suspect that the first and last measurements, ordered by time, are from different timeseries. The right test be to do this assertion within one timeseries only.

Reply:
The `timeseries_name` param doesn't filter for the named timeseries?

https://github.com/oxidecomputer/omicron/blob/bd6c62807fbb2ea4fdae0b11c301124936ea41a2/oximeter/db/src/client.rs#L62-L70

Reply:
I tried it with `criteria = &["name==oximeter"]` and still got the same issue. Here is the response from each of the calls. The criteria look good for each one, but for some reason the combination of `DESC` and `LIMIT 1` gives a data point that's not in the previous queries.

```
[oximeter/db/src/client.rs:1550] all_measurements = Timeseries {
    timeseries_name: "service:request_latency",
    target: Target {
        name: "service",
        fields: [
            Field {
                name: "id",
                value: Uuid(
                    4fa827ea-38bb-c37e-ac2d-f8432ca9c76e,
                ),
            },
            Field {
                name: "name",
                value: String(
                    "oximeter",
                ),
            },
        ],
    },
    metric: Metric {
        name: "request_latency",
        fields: [
            Field {
                name: "method",
                value: String(
                    "POST",
                ),
            },
            Field {
                name: "route",
                value: String(
                    "/b",
                ),
            },
            Field {
                name: "status_code",
                value: I64(
                    204,
                ),
            },
        ],
        datum_type: F64,
    },
    measurements: [
        Measurement {
            timestamp: 2023-06-01T21:59:37.682062Z,
            datum: F64(
                0.0,
            ),
        },
        Measurement {
            timestamp: 2023-06-01T21:59:37.682066Z,
            datum: F64(
                0.0,
            ),
        },
    ],
}
[oximeter/db/src/client.rs:1551] timeseries_asc = Timeseries {
    timeseries_name: "service:request_latency",
    target: Target {
        name: "service",
        fields: [
            Field {
                name: "id",
                value: Uuid(
                    4fa827ea-38bb-c37e-ac2d-f8432ca9c76e,
                ),
            },
            Field {
                name: "name",
                value: String(
                    "oximeter",
                ),
            },
        ],
    },
    metric: Metric {
        name: "request_latency",
        fields: [
            Field {
                name: "method",
                value: String(
                    "POST",
                ),
            },
            Field {
                name: "route",
                value: String(
                    "/b",
                ),
            },
            Field {
                name: "status_code",
                value: I64(
                    204,
                ),
            },
        ],
        datum_type: F64,
    },
    measurements: [
        Measurement {
            timestamp: 2023-06-01T21:59:37.682062Z,
            datum: F64(
                0.0,
            ),
        },
        Measurement {
            timestamp: 2023-06-01T21:59:37.682066Z,
            datum: F64(
                0.0,
            ),
        },
    ],
}
[oximeter/db/src/client.rs:1552] timeseries_desc = Timeseries {
    timeseries_name: "service:request_latency",
    target: Target {
        name: "service",
        fields: [
            Field {
                name: "id",
                value: Uuid(
                    4fa827ea-38bb-c37e-ac2d-f8432ca9c76e,
                ),
            },
            Field {
                name: "name",
                value: String(
                    "oximeter",
                ),
            },
        ],
    },
    metric: Metric {
        name: "request_latency",
        fields: [
            Field {
                name: "method",
                value: String(
                    "POST",
                ),
            },
            Field {
                name: "route",
                value: String(
                    "/b",
                ),
            },
            Field {
                name: "status_code",
                value: I64(
                    204,
                ),
            },
        ],
        datum_type: F64,
    },
    measurements: [
        Measurement {
            timestamp: 2023-06-01T21:59:37.682066Z,
            datum: F64(
                0.0,
            ),
        },
        Measurement {
            timestamp: 2023-06-01T21:59:37.682062Z,
            datum: F64(
                0.0,
            ),
        },
    ],
}
[oximeter/db/src/client.rs:1553] desc_limit_1 = Timeseries {
    timeseries_name: "service:request_latency",
    target: Target {
        name: "service",
        fields: [
            Field {
                name: "id",
                value: Uuid(
                    4fa827ea-38bb-c37e-ac2d-f8432ca9c76e,
                ),
            },
            Field {
                name: "name",
                value: String(
                    "oximeter",
                ),
            },
        ],
    },
    metric: Metric {
        name: "request_latency",
        fields: [
            Field {
                name: "method",
                value: String(
                    "POST",
                ),
            },
            Field {
                name: "route",
                value: String(
                    "/a",
                ),
            },
            Field {
                name: "status_code",
                value: I64(
                    200,
                ),
            },
        ],
        datum_type: F64,
    },
    measurements: [
        Measurement {
            timestamp: 2023-06-01T21:59:37.682007Z,
            datum: F64(
                0.0,
            ),
        },
    ],
}
```

Reply:
The name does filter, yes, but there's more to the timeseries than its name. The name effectively identifies a _schema_, describing the field names and their types (and the measurement type). But each individual timeseries is identified by the actual values of each of those fields. There are as many timeseries with the same name as there are unique combinations of field values.

What you're doing here is basically `SELECT * FROM timeseries_name`. You've selected the "table", and nothing else. Adding criteria, say `id=="some-uuid"`, would be like `SELECT * FROM timeseries_name WHERE id = 'some-uuid'`. Looking at the code for `setup_select_test` might help, or reading some of the summary of RFD 161 as well. Hope that helps.

Reply:
> I tried it with criteria = &["name==oximeter"] and still got the same issue.

Yes, because both timeseries are consistent with this criterion. You'd need to filter on something like the route, say `route==/b`, which only applies to the second timeseries.

Reply:
> I think there's filtering by key that happens at the application level, so things won't work if we're applying the limit in the DB and then filtering

This is why I avoided exposing direct methods for things like `LIMIT` initially. It's difficult to understand the behavior when applying that in a query that itself gives back multiple timeseries. I would probably argue that we either don't support that kind of query right now, or only do so if the timeseries itself is "fully specified", i.e., there's only one timeseries consistent with the provided filtering criteria. There are probably other options too.

Reply:
Whether a given set of criteria determines a single timeseries is a dynamic question, right? i.e., it requires running a query to find out?

Reply:
That's true, yeah. Right now, we issue several queries to the database as part of `select_timeseries_with()`. The second one is to collect the keys that are consistent with the filtering criteria provided to the method. So the timeseries is fully specified iff there is exactly one key returned here. (Zero would also be fine I guess, but that's easy to handle.) So to do what I described above, we could return an error if the limit argument is `Some(_)` and there is more than one key returned from that second query.

That feels a bit janky, but I think it seems better than the confusion you're running into now. WDYT?

Reply:
Yes. I think that's tolerable — at least it avoids broken behavior.

Reply:
Attempted in dc0b5d5b629a649c1738ecd3fafba1e2d6c4e224. It was so easy it makes me nervous.